### PR TITLE
feat(eslint-plugin): [no-shadow] ignoreOnInitialization option

### LINKED ISSUE #4601: [no-shadow] Add `ignoreOnInitialization` option to match the ESLint rule
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-shadow": ["error", { "ignoreOnInitialization": true }],
  }
}
```

```TS
// n/a
```

**Expected Result**

No errors.

**Actual Result**

```
Configuration for rule "@typescript-eslint/no-shadow" is invalid:
Value {"ignoreOnInitialization":true} should NOT have additional properties.
```

**Additional Info**

https://eslint.org/docs/rules/no-shadow#ignoreoninitialization

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.12.1` |
| `@typescript-eslint/parser`        | `5.12.1` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.10.0` |
| `node`                             | `16.14.0` |

### PR DESCRIPTION
**Pull request by Josh-Cena**
## PR Checklist

-   [x] Addresses an existing issue: fixes #4601
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Added an option to line up with base rule

### CODE REVIEW

**Comment by Josh-Cena** [comment]
Not sure if any code should be updated to cope with this option. Seems our scope analyzer is already able to allow this:

```ts
type test = (test: string) => typeof test;
```

Even when `ignoreFunctionTypeParameterNameValueShadow` is turned off

**Review by JoshuaKGoldberg — APPROVED** [review]
Super, thanks! 

@Josh-Cena do you think you have time to add unit tests for this? If not no worries, since it's just a pipe-through of the base rule's config.

**Comment by Josh-Cena** [comment]
Hmmm... No! This rule does not pipe to the base rule but is a re-implementation. We do need to backport the upstream changes and add a test. My bad for not noticing it beforehand

**Comment by Josh-Cena** [comment]
Any reason why this rule does not delegate to the base rule? Should I refactor that in this PR?

**Comment by bradzacher** [comment]
> Any reason why this rule does not delegate to the base rule?

At the end of the AST the base rule iterates over every single scope and variable using the scope analysis APIs and then individually reports on variables that match its internal logic.

We unfortunately can't pick-and-choose which variables the base rule might report on unless we hackily override the scope analysis APIs - which means we can't delegate! Thus we have to maintain a complete fork, sadly.

If you can figure out a clean way to refactor this to reuse the logic - feel free! I don't think that it's possible to do though.

**Comment by Josh-Cena** [comment]
Ugh, I see. That's unfortunate. I'll incorporate upstream changes later today.

**Comment by Josh-Cena** [comment]
The unit tests seem to indicate that our extension rule is reporting more errors than the base rule on the same case. Is that because of test setup, or because of our scope manager having different logics?

File: packages/eslint-plugin/src/rules/no-shadow.ts [hunk 1]
```diff
@@ -24,6 +24,9 @@ type Options = [
   },
 ];
 
+const SENTINEL_TYPE =
+  /^(?:(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression|CatchClause|ImportDeclaration|ExportNamedDeclaration)$/u;
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/no-shadow.ts:28]: *personally* I hate this style of check.
This is so opaque and complicated to parse and understand.

```suggestion
const SENTINEL_TYPE: ReadonlySet<AST_NODE_TYPE> = new Set([
  AST_NODE_TYPE.ArrowFunctionExpression,
  AST_NODE_TYPE.CatchClause,
  AST_NODE_TYPE.ClassDeclaration,
  AST_NODE_TYPE.ClassExpression,
  AST_NODE_TYPE.ExportNamedDeclaration,
  AST_NODE_TYPE.FunctionDeclaration,
  AST_NODE_TYPE.FunctionExpression,
  AST_NODE_TYPE.ImportDeclaration,
]);
```

**Comment by lopezjurip** [comment]
any progress on this? or a possible workaround?

**Comment by bradzacher** [comment]
@lopezjurip - please don't comment on issues and prs asking for progress.
The progress as seen in this PR.
The volunteer contributors will progress it forward as they find the time.
Commenting asking for updates just spams everyone with an unnecessary notification.

**Comment by Josh-Cena** [comment]
Still quite clueless. If I copied the test cases from upstream, but our rule testers report 3 errors while the original test case only expects one, how could I get the data about the extra errors? How do people usually debug this?

**Comment by bradzacher** [comment]
> How do people usually debug this?

you can use the vscode debugger to debug the tests - we have it configured so you just need to open the file and select the lint rule debugger.

also you can set `only: true` on the specific test case you're looking at so you can focus and remove noise.

**Comment by Josh-Cena** [comment]
I finally got time to look at this. I had to hack into eslint's rule tester so I can let it log the actual line/column of the report. Turns out that the tests on their repo would turn all source code into one line, so the line/column numbers are different from ours.

**Pull request merged**